src/rgw/rgw_op.cc

```diff
@@ -910,6 +911,27 @@ void rgw_build_iam_environment(rgw::sal::Driver* driver,
   }
 }
 
+void handle_replication_status_header(
+    const DoutPrefixProvider *dpp,
+    rgw::sal::Attrs& attrs,
+    req_state* s,
+    const ceph::real_time &obj_mtime) {
+  auto attr_iter = attrs.find(RGW_ATTR_OBJ_REPLICATION_STATUS);
+  if (attr_iter != attrs.end() && attr_iter->second.to_str() == "PENDING") {
+    if (s->object->is_sync_completed(dpp, obj_mtime)) {
+        s->object->set_atomic();
+        rgw::sal::Attrs setattrs, rmattrs;
+        bufferlist bl;
+        bl.append("COMPLETED");
+        setattrs[RGW_ATTR_OBJ_REPLICATION_STATUS] = std::move(bl);
+	int ret = s->object->set_obj_attrs(dpp, &setattrs, &rmattrs, s->yield, 0);
```

Comment:
cc @smanjara @thotz 

i've been seeing multisite failures recently, and this seems to be the culprit. the underlying bug is https://tracker.ceph.com/issues/64173, whose fix in https://github.com/ceph/ceph/pull/54393 still hasn't merged

to summarize, this call to `set_obj_attrs()` does a bucket index transaction, but passes empty values for the owner, etag, storage class, etc. this causes ListObjects to return empty values for Owner and ETag, leading to test failures like this:
```
======================================================================
FAIL: rgw_multi.tests.test_bucket_reshard_incremental
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_teuthology_454a46106b61e201598ae5e1c392af2328539bfe/virtualenv/lib/python3.10/site-packages/nose/case.py", line 170, in runTest
    self.test(*self.arg)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_532ed64a904276d871b91839ffec7885dab6b986/qa/../src/test/rgw/rgw_multi/tests.py", line 1692, in test_bucket_reshard_incremental
    zonegroup_bucket_checkpoint(zonegroup_conns, bucket.name)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_532ed64a904276d871b91839ffec7885dab6b986/qa/../src/test/rgw/rgw_multi/tests.py", line 401, in zonegroup_bucket_checkpoint
    target_conn.check_bucket_eq(source_conn, bucket_name)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_532ed64a904276d871b91839ffec7885dab6b986/qa/tasks/rgw_multi/zone_rados.py", line 84, in check_bucket_eq
    check_object_eq(k1, k2)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_532ed64a904276d871b91839ffec7885dab6b986/qa/tasks/rgw_multi/zone_rados.py", line 34, in check_object_eq
    eq(k1.owner.id, k2.owner.id)
AssertionError: 'RGW11111111111111111' != ''
```
when comparing the output of `radosgw-admin bucket list` and `object stat`, i see that objects with "user.rgw.amz-replication-status": "COMPLETE" have empty values in the index:
```
        "meta": {
            "category": 1,
            "size": 3,
            "mtime": "2024-07-11T02:44:52.655546Z",
            "etag": "",
            "storage_class": "",
            "owner": "",
            "owner_display_name": "",
            "content_type": "",
```
while the entries are still intact for objects with "user.rgw.amz-replication-status": "PENDING":
```
        "meta": {
            "category": 1,
            "size": 3,
            "mtime": "2024-07-11T02:44:52.610456Z",
            "etag": "acbd18db4cc2f85cedef654fccc4a4d8",
            "storage_class": "",
            "owner": "RGW11111111111111111",
            "owner_display_name": "",
            "content_type": "application/octet-stream",
```

Reply:
Can we work to merge the PR that fixes the underlying bug? Anything I can do to help out?

Reply:
thanks @awojno-bloomberg. i'm trying to get this unblocked in https://github.com/ceph/ceph/pull/54393#pullrequestreview-2172100453. you might help by testing that out to verify that it fixes the transition to COMPLETED?

Reply:
fix merged in https://github.com/ceph/ceph/pull/58649